### ISSUE TITLE
Create `distillation/loss.py`

### ISSUE THREAD

**Issue by peremartra** [peremartra/optipfair#23]
# Sub-issue 2/5: Create `distillation/loss.py`

**Parent issue:** #21 
**Depends on:** Sub-issue 1 (mapping module must exist)

---

## Context

This task implements the core loss function for knowledge distillation in OptiPFair. It is a direct adaptation of `compute_compound_loss_advanced()` from `CH06_NB03_Hands_on_qwen3_5.ipynb`, renamed and integrated into OptiPFair conventions.

The function combines up to four loss components:

| # | Name | Method | Controlled by |
|---|------|--------|---------------|
| 1 | Task Loss | Cross-Entropy with hard labels | `alpha` |
| 2 | Logits Loss | Skew KLD — interpolates Forward/Reverse KLD (DistiLLM-2) | `beta` |
| 3 | Trajectory Loss | Cosine similarity between hidden states (feature alignment) | `gamma` |
| 4 | FDD Derivative Loss | Cosine similarity of deltas between consecutive blocks (ACL 2025) | `delta` |

**Default values are calibrated for OptiPFair's target use case: small models and small datasets.**

| Parameter | Default | Rationale |
|---|---|---|
| `alpha` | `0.6` | Higher hard label weight suits small datasets. Validated in NB03. |
| `beta` | `0.4` | `alpha + beta = 1.0` when features are off. |
| `gamma` | `0.0` | Feature alignment off by default. |
| `delta` | `0.0` | Feature alignment off by default. |
| `temperature` | `2.0` | Standard KD value. |
| `skew_alpha` | `0.4` | 60% teacher + 40% student. Validated in NB03. Recommended range for small models. |

---

## Files to Create
```
optipfair/distillation/loss.py          # the implementation
tests/test_distillation_loss.py         # tests
```

---

## Reference Implementation

The following is the original function from `CH06_NB03_Hands_on_qwen3_5.ipynb`. The OptiPFair version (`compute_distillation_loss`) must replicate this logic exactly, only changing the function name and default parameter values.
```python
def compute_compound_loss_advanced(
    student_logits,      # [batch, seq_len, vocab_size]
    teacher_logits,      # [batch, seq_len, vocab_size]
    student_hiddens,     # List of [batch, seq_len, hidden_dim] or None
    teacher_hiddens,     # List of [batch, seq_len, hidden_dim] or None
    labels,              # [batch, seq_len]
    layer_map=None,
    alpha=0.1,
    beta=0.8,
    gamma=0.1,
    delta=0.1,
    temperature=2.0,
    skew_alpha=0.1
):
    device = student_logits.device

    # 1. TASK LOSS
    shift_logits = student_logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()
    loss_task = F.cross_entropy(
        shift_logits.view(-1, shift_logits.size(-1)),
        shift_labels.view(-1),
        ignore_index=-100
    )

    # 2. SKEW KLD
    with torch.no_grad():
        student_probs = F.softmax(student_logits[..., :-1, :] / temperature, dim=-1)
        teacher_probs = F.softmax(teacher_logits[..., :-1, :] / temperature, dim=-1)
        mixed_probs = skew_alpha * student_probs + (1 - skew_alpha) * teacher_probs

    student_log_probs = F.log_softmax(student_logits[..., :-1, :] / temperature, dim=-1)
    kl_elementwise = student_probs * (student_log_probs - torch.log(mixed_probs + 1e-9))
    loss_logits = kl_elementwise.sum(dim=-1).mean() * (temperature ** 2)

    # 3. TRAJECTORY LOSS
    if gamma > 0 and student_hiddens is not None and teacher_hiddens is not None:
        loss_trajectory = 0.0
        for student_idx, teacher_idx in enumerate(layer_map):
            student_h = student_hiddens[student_idx]
            teacher_h = teacher_hiddens[teacher_idx]
            student_flat = student_h.reshape(-1, student_h.size(-1))
            teacher_flat = teacher_h.reshape(-1, teacher_h.size(-1))
            student_norm = F.normalize(student_flat, p=2, dim=1)
            teacher_norm = F.normalize(teacher_flat, p=2, dim=1)
            cos_sim = (student_norm * teacher_norm).sum(dim=1).mean()
            loss_trajectory += (1 - cos_sim)
        loss_trajectory = loss_trajectory / len(layer_map)
    else:
        loss_trajectory = torch.tensor(0.0, device=device)

    # 4. FDD DERIVATIVE LOSS
    loss_derivative = torch.tensor(0.0, device=device)
    if delta > 0 and student_hiddens is not None and teacher_hiddens is not None:
        num_derivatives = 0
        for student_idx in range(len(layer_map) - 1):
            teacher_idx = layer_map[student_idx]
            teacher_idx_next = layer_map[student_idx + 1]
            student_delta = student_hiddens[student_idx + 1] - student_hiddens[student_idx]
            teacher_delta = teacher_hiddens[teacher_idx_next] - teacher_hiddens[teacher_idx]
            student_delta_flat = student_delta.reshape(-1, student_delta.size(-1))
            teacher_delta_flat = teacher_delta.reshape(-1, teacher_delta.size(-1))
            student_delta_norm = F.normalize(student_delta_flat, p=2, dim=1)
            teacher_delta_norm = F.normalize(teacher_delta_flat, p=2, dim=1)
            cos_sim_deriv = (student_delta_norm * teacher_delta_norm).sum(dim=1).mean()
            loss_derivative += (1 - cos_sim_deriv)
            num_derivatives += 1
        if num_derivatives > 0:
            loss_derivative = loss_derivative / num_derivatives

    # COMBINE
    total_loss = (
        alpha * loss_task +
        beta * loss_logits +
        gamma * loss_trajectory +
        delta * loss_derivative
    )

    loss_dict = {
        'total': total_loss.item(),
        'task': loss_task.item(),
        'logits': loss_logits.item(),
        'trajectory': loss_trajectory.item(),
        'derivative': loss_derivative.item()
    }

    return total_loss, loss_dict
```

---

### `optipfair/distillation/loss.py`

Adapt the reference implementation above as `compute_distillation_loss()` with:
- Updated default values (`alpha=0.6`, `beta=0.4`, `gamma=0.0`, `delta=0.0`, `skew_alpha=0.4`)
- Updated parameter comments:
```python
alpha: float = 0.6,    # Hard labels weight: learns from ground truth (dataset)
beta: float = 0.4,     # Soft labels weight: learns from teacher's output distribution
gamma: float = 0.0,    # Feature alignment weight: matches hidden state representations layer-by-layer
delta: float = 0.0,    # Feature dynamics weight: matches how representations change between layers
```
- Full docstring referencing DistiLLM-2 (2024) and FDD ACL 2025
- Return type annotation: `Tuple[torch.Tensor, Dict[str, float]]`

---

## Tests

Create `tests/test_distillation_loss.py` using synthetic tensors (no real model needed):
```python
"""
Tests for the distillation loss module.
"""

import unittest
import torch
import sys
import os

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from optipfair.distillation.loss import compute_distillation_loss


class TestComputeDistillationLoss(unittest.TestCase):

    def setUp(self):
        """Create minimal synthetic tensors for testing."""
        self.batch, self.seq, self.vocab, self.hidden = 2, 8, 100, 32
        self.n_layers = 4

        self.student_logits = torch.randn(self.batch, self.seq, self.vocab)
        self.teacher_logits = torch.randn(self.batch, self.seq, self.vocab)
        self.labels = torch.randint(0, self.vocab, (self.batch, self.seq))

        self.student_hiddens = [
            torch.randn(self.batch, self.seq, self.hidden)
            for _ in range(self.n_layers)
        ]
        self.teacher_hiddens = [
            torch.randn(self.batch, self.seq, self.hidden)
            for _ in range(self.n_layers)
        ]
        self.layer_map = list(range(self.n_layers))

    def test_returns_tuple(self):
        loss, loss_dict = compute_distillation_loss(
            self.student_logits, self.teacher_logits,
            None, None, self.labels
        )
        self.assertIsInstance(loss, torch.Tensor)
        self.assertIsInstance(loss_dict, dict)

    def test_loss_dict_always_has_five_keys(self):
        _, loss_dict = compute_distillation_loss(
            self.student_logits, self.teacher_logits,
            None, None, self.labels
        )
        self.assertEqual(
            set(loss_dict.keys()),
            {'total', 'task', 'logits', 'trajectory', 'derivative'}
        )

    def test_disabled_components_are_zero(self):
        """With gamma=0 and delta=0, trajectory and derivative must be 0.0."""
        _, loss_dict = compute_distillation_loss(
            self.student_logits, self.teacher_logits,
            None, None, self.labels,
            gamma=0.0, delta=0.0
        )
        self.assertEqual(loss_dict['trajectory'], 0.0)
        self.assertEqual(loss_dict['derivative'], 0.0)

    def test_total_loss_is_tensor_with_grad(self):
        self.student_logits.requires_grad_(True)
        loss, _ = compute_distillation_loss(
            self.student_logits, self.teacher_logits,
            None, None, self.labels
        )
        self.assertTrue(loss.requires_grad)

    def test_feature_alignment_components_nonzero(self):
        """With gamma>0 and delta>0 and valid hidden states, components must be nonzero."""
        _, loss_dict = compute_distillation_loss(
            self.student_logits, self.teacher_logits,
            self.student_hiddens, self.teacher_hiddens, self.labels,
            layer_map=self.layer_map,
            gamma=0.1, delta=0.1
        )
        self.assertGreater(loss_dict['trajectory'], 0.0)
        self.assertGreater(loss_dict['derivative'], 0.0)

    def test_loss_dict_values_are_floats(self):
        _, loss_dict = compute_distillation_loss(
            self.student_logits, self.teacher_logits,
            None, None, self.labels
        )
        for key, value in loss_dict.items():
            self.assertIsInstance(value, float, f"Key '{key}' is not a float")

    def test_default_parameters_sum_to_one(self):
        """Default alpha + beta must equal 1.0 (features off by default)."""
        import inspect
        sig = inspect.signature(compute_distillation_loss)
        alpha = sig.parameters['alpha'].default
        beta = sig.parameters['beta'].default
        gamma = sig.parameters['gamma'].default
        delta = sig.parameters['delta'].default
        self.assertEqual(gamma, 0.0)
        self.assertEqual(delta, 0.0)
        self.assertAlmostEqual(alpha + beta, 1.0, places=6)


if __name__ == "__main__":
    unittest.main()
```

---

## Acceptance Criteria

- [ ] `optipfair/distillation/loss.py` exists with `compute_distillation_loss()`
- [ ] Default values: `alpha=0.6`, `beta=0.4`, `gamma=0.0`, `delta=0.0`, `temperature=2.0`, `skew_alpha=0.4`
- [ ] `loss_dict` always contains exactly five keys: `total`, `task`, `logits`, `trajectory`, `derivative`
- [ ] `trajectory` and `derivative` are `0.0` when `gamma=0` and `delta=0`
- [ ] The returned loss tensor has `requires_grad=True` (backpropagation works)
- [ ] All tests in `tests/test_distillation_loss.py` pass
